### PR TITLE
feat: Feat/hub improvement

### LINKED ISSUE #196: Add a chevron to collapse dashboards under an organisation
- [ ] Add a chevron button to collapse dashboards under an organisation
image

### LINKED ISSUE #199: Pin the collection list to the right. And scroll the dashbords to the left independently

### LINKED ISSUE #202: Full text search on dashboard titles in ARLAS Hub
### Description of the current behavior

Currently, it is possible in ARLAS Hub to filter dashboards by collections, but it is not possible to filter dashboards by name to quickly retrieve it. Also, it is tricky to share a link to a dashboard if its id changes.

### Description of the desired behavior

To have a full text search in the ARLAS Hub to filter the dashboards: only the one having the keywords in their titles are shown. The keywords are in the url so that the url can be shared with others: when they open the link, the filtering is applied.

### PR DESCRIPTION
**Pull request by MAudelGisaia**
- close #202
- close #199 
- close #196 
- 
[Capture vidéo du 07-11-2024 16:10:30.webm](https://github.com/user-attachments/assets/da78d4d8-2c24-416f-956a-b96fae39c2f5)

image

### CODE REVIEW

File: src/app/components/dynamic-hub/collapse/card-dropdown.component.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<button class="dynamic-collapse" mat-icon-button (click)="collapse()">
+    <mat-icon *ngIf="display; else arrow_down">keyboard_arrow_up</mat-icon>
+    <ng-template  #arrow_down>
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/collapse/card-dropdown.component.html:3]: extra space after ng-template

File: src/app/components/dynamic-hub/collapse/card-dropdown.component.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { Component, Input, Renderer2 } from '@angular/core';
+
+@Component({
+    selector: '[arlas-card-dropdown]',
+    templateUrl: './card-dropdown.component.html',
+    styles: ['.dynamic-collapse{height: 30px;width: 30px;line-height: 30px;}'],
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/collapse/card-dropdown.component.ts:24]: Why not create a scss file like for other components?

File: src/app/components/dynamic-hub/dynamic-hub.component.html [hunk 3]
```diff
@@ -41,15 +42,18 @@ <h2>
...
-                    [class.disabled]="!col.value.selected" (click)="getCheckbox(!col.value.selected, col.key)"
-                    *ngFor="let col of cardCollections | keyvalue">
-                    <mat-card-header>
-                        {{col.key | getCollectionDisplayName | translate}}
-                    </mat-card-header>
-                </mat-card>
+            <div class="collection-wrapper">
+                <h2>{{'Collections' | translate}}</h2>
+                <div class="cards">
+                    <mat-card class="card" [style.borderLeftColor]="col.value.color" style="border-left: 3px solid;"
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.html:48]: Can the style be moved to the `card` class?

File: src/app/components/dynamic-hub/dynamic-hub.component.scss [hunk 4]
```diff
@@ -76,8 +86,33 @@ h2 {
         .collection {
             flex: 1;
 
+            &-wrapper {
+                height: calc(100vh - 60px);
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.scss:90]: why 60px?

File: src/app/components/dynamic-hub/dynamic-hub.component.scss [hunk 5]
```diff
@@ -76,8 +86,33 @@ h2 {
...
+                position: sticky;
+                width: 100%;
+                top: $header-height;
+            }
+
             .cards {
                 padding-top: 1px;
+                height: calc(100% - 90px);
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.scss:98]: why 90px?

File: src/app/components/dynamic-hub/dynamic-hub.component.scss [hunk 6]
```diff
@@ -76,8 +86,33 @@ h2 {
...
+                overflow-x: hidden;
+                overflow-y: scroll;
+                & .scroll-hider {
+                    position: absolute;
+                    background: #F5F5F5;
+                    height: 100%;
+                    top: 0;
+                    right: 0;
+                    width: .4em;
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.scss:107]: I think we are trying to phase out of em sizes to just use px

File: src/app/components/dynamic-hub/dynamic-hub.component.ts [hunk 7]
```diff
@@ -319,6 +323,13 @@ export class DynamicHubComponent implements OnInit {
         if (!this.cardCollections.has(c.collection)) {
             this.cardCollections.set(c.collection, { color: c.color, selected: true });
         }
+        // todo: remove when dev end
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.ts:326]: To remove later

File: src/app/components/dynamic-hub/dynamic-hub.component.ts [hunk 8]
```diff
@@ -366,6 +377,62 @@ export class DynamicHubComponent implements OnInit {
         } else {
             this.cards = this.cardsRef;
         }
+
+        // todo: remove when dev end
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.ts:381]: To remove later

File: src/app/components/dynamic-hub/dynamic-hub.component.ts [hunk 9]
```diff
@@ -366,6 +377,62 @@ export class DynamicHubComponent implements OnInit {
...
+    public filterDashboard(searchValue: string){
+        if(searchValue){
+            this.cardsFiltered = new Map<string, Card[]>();
+            this.searchIndex.filter(s => s.split(';search:')[1].includes(searchValue.toLowerCase().trim())).forEach(f => {
+                const z = f.split(';');
+                const key = z[0].split(':')[1];
+                const index = z[1].split(':')[1];
+
+                if(this.cardsFiltered.has(key)){
+                    this.cardsFiltered.get(key).push( this.cards.get(key)[index]);
```
Comment (QuCMGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.ts:427]: extra space after push(

**Review by QuCMGisaia — COMMENTED** [review]
I know it is not the purpose of the PR, but would it be possible for you to add some padding around the create and import dashboards button so it can look a bit better? :pray: 

It looks nice with the collections like that to the side !!

File: src/app/components/dynamic-hub/collapse/card-dropdown.component.ts (hunk 2 above)
Comment (MAudelGisaia) [reply src/app/components/dynamic-hub/collapse/card-dropdown.component.ts:24]: do not need for 3 line of scss that ll never change.

File: src/app/components/dashboard-search/dashboard-search.component.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+@Component({
+    selector: 'arlas-dashboard-search',
+    templateUrl: './dashboard-search.component.html',
+    styleUrls: ['./dashboard-search.component.scss']
+})
+export class DashboardSearchComponent implements OnInit{
+
+    /**
+     * @Input : Angular
```
Comment (QuCMGisaia) [inline src/app/components/dashboard-search/dashboard-search.component.ts:34]: Description does not match

File: src/assets/i18n/en.json [hunk 11]
```diff
@@ -59,5 +59,6 @@
 	"Missing permissions to create a dashboard": "Missing permissions to create or update a dashboard",
 	"no_organisation":"",
 	"Choose an organisation": "Choose an organisation",
-	"Select": "Select"
+	"Select": "Select",
+    "searchPlaceholder": "Search..."
```
Comment (QuCMGisaia) [inline src/assets/i18n/en.json:63]: Search a dashboard...

File: src/assets/i18n/fr.json [hunk 12]
```diff
@@ -60,6 +60,6 @@
 	"Missing permissions to create a dashboard": "Vous n'avez pas les droits suffisants pour créer ou modifier un dashboard",
 	"no_organisation":"",
 	"Choose an organisation": "Choisissez une organisation",
-	"Select": "Sélectionner"
-
+	"Select": "Sélectionner",
+    "searchPlaceholder": "Rechercher..."
```
Comment (QuCMGisaia) [inline src/assets/i18n/fr.json:64]: Rechercher un dashboard...

**Review by QuCMGisaia — APPROVED** [review]

File: src/app/components/dynamic-hub/dynamic-hub.component.html (hunk 3 above)
Comment (MohamedHamouGisaia) [reply src/app/components/dynamic-hub/dynamic-hub.component.html:48]: It is possible for the `style="border-left: 3px solid;"`
but not the `[style.borderLeftColor]="col.value.color"`

File: src/app/components/dynamic-hub/dynamic-hub.component.ts [hunk 13]
```diff
@@ -366,6 +377,62 @@ export class DynamicHubComponent implements OnInit {
...
+        const queryString = window.location.search;
+        const urlParams = new URLSearchParams(queryString);
+        const filter = urlParams.get('filter_hub_dashboard');
+        this.filterDashboard(filter);
+    }
+
+    private _buildSearchIndex(){
+        this.cards.forEach((values: Card[], key: string) => {
+            values.forEach((c,i) => {
+                const s = `${c.title.toLowerCase().replace(/\s/g, '')}${(c?.organisation) ? c.organisation : '' }`;
```
Comment (MohamedHamouGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.ts]: Please preceed (c?.organisation) by !! in order to make it clear that we check for existence

File: src/app/components/dynamic-hub/dynamic-hub.component.ts (hunk 9 above)
Comment (MohamedHamouGisaia) [reply src/app/components/dynamic-hub/dynamic-hub.component.ts]: could you add a check if **this.cards.get(key)** exists ?
It is possible that the `this.cards` map gets reset by a refresh call and be for an instant empty

File: src/app/services/dashboard-search.service.ts [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        } else {
+            urlParams.delete(this.URL_PARAMS);
+        }
+        window.history.replaceState(null, '', query);
+    }
+
+    public retrieveParamsFromUrl(){
+        const queryString = window.location.search;
+        const urlParams = new URLSearchParams(queryString);
+        this._currentValue = urlParams.get(this.URL_PARAMS);
```
Comment (MohamedHamouGisaia) [inline src/app/services/dashboard-search.service.ts:74]: Does urlParams decode the url ?

File: src/app/components/dynamic-hub/dynamic-hub.component.ts [hunk 15]
```diff
@@ -366,6 +387,34 @@ export class DynamicHubComponent implements OnInit {
...
+    public filterDashboard(searchValue: string, preserveEmptyCardList = false){
+        let previous;
+        if(searchValue){
+            previous = this.cardsFiltered;
+            this.cardsFiltered = new Map<string, Card[]>();
+            this.dashboardSearch
+                .getMatchingSearchIndexes()
+                .forEach(searchIndex => {
+                    if(this.cardsFiltered.has(searchIndex.key)){
+                        this.cardsFiltered.get(searchIndex.key).push(this.cards.get(searchIndex.key)[searchIndex.cardIndex]);
```
Comment (MohamedHamouGisaia) [inline src/app/components/dynamic-hub/dynamic-hub.component.ts:404]: could you add a check if **this.cards.get(key)** exists ?
It is possible that the `this.cards` map gets reset by a refresh call and be for an instant empty

Comment (MAudelGisaia) [reply src/app/components/dynamic-hub/dynamic-hub.component.ts:404]: a refresh triggered by a button ?

File: src/app/services/dashboard-search.service.ts (hunk 14 above)
Comment (MAudelGisaia) [reply src/app/services/dashboard-search.service.ts:74]: yes, its decode the search

File: src/app/components/dynamic-hub/dynamic-hub.component.ts [hunk 16]
```diff
@@ -378,61 +388,33 @@ export class DynamicHubComponent implements OnInit {
...
-
-                if(this.cardsFiltered.has(key)){
-                    this.cardsFiltered.get(key).push( this.cards.get(key)[index]);
-                } else {
-                    this.cardsFiltered.set(key, [this.cards.get(key)[index]]);
-                }
-
-            });
+            this.dashboardSearch
+                .getMatchingSearchIndexes()
```
[the remark this answers is not in the document]
Comment (MAudelGisaia) [reply src/app/components/dynamic-hub/dynamic-hub.component.ts:401]: ok

**Pull request merged**